routing/control_tower.go

```diff
@@ -85,10 +132,108 @@ func (p *controlTower) Success(paymentHash lntypes.Hash,
 func (p *controlTower) Fail(paymentHash lntypes.Hash,
 	reason channeldb.FailureReason) error {
 
-	return p.db.Fail(paymentHash, reason)
+	err := p.db.Fail(paymentHash, reason)
+	if err != nil {
+		return err
+	}
+
+	// Notify subscribers of fail event.
+	p.notifyFinalEvent(
+		paymentHash, PaymentResult{
+			Success:       false,
+			FailureReason: reason,
+		},
+	)
+
+	return nil
 }
 
 // FetchInFlightPayments returns all payments with status InFlight.
 func (p *controlTower) FetchInFlightPayments() ([]*channeldb.InFlightPayment, error) {
 	return p.db.FetchInFlightPayments()
 }
+
+// SubscribePayment subscribes to updates for the payment with the given hash.
+// It returns a boolean indicating whether the payment is still in flight and a
+// channel that provides the final outcome of the payment.
+func (p *controlTower) SubscribePayment(paymentHash lntypes.Hash) (
+	bool, chan PaymentResult, error) {
+
+	// Create a channel with buffer size 1. For every payment there will be
+	// exactly one event sent.
+	c := make(chan PaymentResult, 1)
+
+	// Take lock before querying the db to prevent this scenario:
+	// FetchPayment returns us an in-flight state -> payment succeeds, but
+	// there is no subscriber to notify yet -> we add ourselves as a
+	// subscriber -> ... we will never receive a notification.
+	p.subscribersMtx.Lock()
+	defer p.subscribersMtx.Unlock()
+
+	payment, err := p.db.FetchPayment(paymentHash)
+	if err != nil {
+		return false, nil, err
+	}
+
+	var event PaymentResult
+
+	switch payment.Status {
+
+	// Payment is currently in flight. Register this subscriber and
+	// return without writing a result to the channel yet.
+	case channeldb.StatusInFlight:
+		p.subscribers[paymentHash] = append(
+			p.subscribers[paymentHash], c,
+		)
+
+		return true, c, nil
```

Comment:
Ok, this is much better now that we don't have to return this from the DB.

Reply:
It isn't new, I just restored code that I had previously. Do you mean it is better because the bool isn't set in the db tx?